stumpy/core.py

```diff
@@ -1926,3 +1926,140 @@ def _idx_to_mp(I, T, m, normalize=True):
+        The third row represents diagonal indices to traverse in the tile
+        The fourth row represents weight and validity of the tile
+    """
+    # dimensions of distance matrix
+    dim = (n_A - m + 1, n_B - m + 1)
+    # number of tiles
+    tiles_count = np.uint64(np.ceil(dim[0] / tile_size) * np.ceil(dim[1] / tile_size))
+    tiles = np.zeros((tiles_count, 4, 2), dtype=np.int64)
+
+    tile_idx = 0
+    # iterate over each tile
+    for y_offset in range(0, dim[0], tile_size):
+        # height of current tile
+        tile_height = min(tile_size, dim[0] - y_offset)
+        for x_offset in range(0, dim[1], tile_size):
+            # width of current tile
+            tile_width = min(tile_size, dim[1] - x_offset)
+
+            # lower and upper diagonal indices to traverse within tile
+            tiles_lower_diag = max(1 - tile_height, diags[0] + y_offset - x_offset)
+            tiles_upper_diag = min(tile_width, diags[-1] + 1 + y_offset - x_offset)
+
+            # count number of computable distances
+            if tiles_lower_diag > 0:
+                tile_weight = 0
+                computable_diags = tile_width - tiles_lower_diag
+                tile_weight += (computable_diags * (computable_diags + 1)) // 2
+
+                ignorable_diags = tile_width - tiles_upper_diag - 1
+                tile_weight -= (ignorable_diags * (ignorable_diags + 1)) // 2
+            elif tiles_upper_diag > 0:
+                tile_weight = tile_width * tile_height
+                ignorable_diags = tile_height + tiles_lower_diag - 1
+                tile_weight -= (ignorable_diags * (ignorable_diags + 1)) // 2
+
+                ignorable_diags = tile_width - tiles_upper_diag - 1
+                tile_weight -= (ignorable_diags * (ignorable_diags + 1)) // 2
+            else:
+                tile_weight = 0
+                computable_diags = tile_height + tiles_upper_diag
+                tile_weight += (computable_diags * (computable_diags + 1)) // 2
+
+                ignorable_diags = tile_height + tiles_lower_diag - 1
+                tile_weight -= (ignorable_diags * (ignorable_diags + 1)) // 2
+
+            tile_validity = 0 if tiles_lower_diag >= tiles_upper_diag else 1
+
+            # offset position of tile
+            tiles[tile_idx][0][0] = y_offset
+            tiles[tile_idx][0][1] = x_offset
+            # dimensions of tile
+            tiles[tile_idx][1][0] = tile_height
+            tiles[tile_idx][1][1] = tile_width
+            # diagonals of tile to traverse
+            tiles[tile_idx][2][0] = tiles_lower_diag
+            tiles[tile_idx][2][1] = tiles_upper_diag
+            # weight of tile
+            tiles[tile_idx][3][0] = tile_weight
+            # validity of tile (i.e. whether or not to skip this tile)
+            tiles[tile_idx][3][1] = tile_validity
+
+            tile_idx += 1
+
+    return tiles
+
+
+@njit()
+def _get_tiles_ranges(tiles, n_threads):
```

Comment:
Done! I changed it to use `_get_array_ranges` instead. I set the `truncate` parameter to `True`, but it probably doesn't matter, right?

Reply:
Never mind, I think the tests are failing because I set it to `True`

Reply:
> Never mind, I think the tests are failing because I set it to True

Yeah, there are only (very) rare circumstances where `truncate` should be set to `True` and this is likely not one of those cases. In this case, I think you end up with threads with no start/stop range (i.e., there is a mismatch between the number of start/stop indices and the number of threads) and the function fails to execute. When you leave `truncate=False`, even if the number of threads is less than the amount of work/diagonals available, the additional threads will still be assigned a (fake/no-op) start/stop range except that it just does zero work. Let's see if your setting `truncate=False` resolves the problem.

Reply:
Awesome! Looks like the tests are passing again

Reply:
Yepp, my bad, you're right I was getting "index out of bounds" errors because the tile indices were "truncated" instead of "faked".

There are still a few more things to refactor, but I think overall it's coming to place. Lemme know if you have any further feedback.